src/v/cluster/rm_partition_frontend.cc

```diff
@@ -60,28 +70,111 @@ ss::future<begin_tx_reply> rm_partition_frontend::begin_tx(
+    if (!has_metadata) {
+        vlog(clusterlog.warn, "can't find {} in the metadata cache", ntp);
+        co_return begin_tx_reply{
+          .ntp = ntp, .ec = tx_errc::partition_not_exists};
     }
 
-    auto _self = _controller->self();
-
-    if (leader == _self) {
-        return do_begin_tx(ntp, pid, tx_seq, transaction_timeout_ms);
+    retries = _metadata_dissemination_retries;
+    aborted = false;
+    auto leader_opt = _leaders.local().get_leader(ntp);
+    while (!aborted && !leader_opt && 0 < retries--) {
+        vlog(
+          clusterlog.warn,
+          "waiting for {} to fill leaders cache, retries left: {}",
+          ntp,
+          retries);
+        aborted = !co_await sleep_abortable(delay_ms);
+        leader_opt = _leaders.local().get_leader(ntp);
     }
 
-    vlog(clusterlog.trace, "dispatching begin tx to {} from {}", leader, _self);
+    retries = _metadata_dissemination_retries;
+    delay_ms = _metadata_dissemination_retry_delay_ms;
+    aborted = false;
+    while (!aborted && 0 < retries--) {
+        if (!leader_opt) {
+            vlog(clusterlog.warn, "can't find {} in the leaders cache", ntp);
+            co_return begin_tx_reply{
+              .ntp = ntp, .ec = tx_errc::leader_not_found};
+        }
+
+        auto leader = leader_opt.value();
+        auto _self = _controller->self();
+
+        begin_tx_reply result;
+        if (leader == _self) {
+            result = co_await do_begin_tx(
+              ntp, pid, tx_seq, transaction_timeout_ms);
+            if (result.ec == tx_errc::leader_not_found) {
+                vlog(
+                  clusterlog.warn,
+                  "local execution of begin_tx({},...) failed with 'not a "
+                  "leader', retries left: {}",
+                  ntp,
+                  retries);
+                aborted = !co_await sleep_abortable(delay_ms);
+                leader_opt = _leaders.local().get_leader(ntp);
+                continue;
+            }
+            if (result.ec != tx_errc::none) {
+                vlog(
+                  clusterlog.warn,
+                  "local execution of begin_tx({},...) failed with {}",
+                  ntp,
+                  result.ec);
+            }
+            co_return result;
+        }
+
+        vlog(
+          clusterlog.trace,
+          "dispatching begin_tx({},...) from {} to {}",
+          ntp,
+          _self,
+          leader);
+        result = co_await dispatch_begin_tx(
+          leader, ntp, pid, tx_seq, transaction_timeout_ms, timeout);
+        if (result.ec == tx_errc::leader_not_found) {
+            vlog(
+              clusterlog.warn,
+              "remote execution of begin_tx({},...) on {} failed with 'not a "
+              "leader', retries left: {}",
+              ntp,
+              leader,
+              retries);
+            aborted = !co_await sleep_abortable(delay_ms);
+            leader_opt = _leaders.local().get_leader(ntp);
```

Comment:
same here. it seems like this is a retry loop, but if leader_opt is null-opt, then right after continue an error is returned without retrying.